### PR TITLE
Spark 3.5: Add validation to SparkConfParser

### PR DESCRIPTION
**Pull request by aokolnychyi**
This PR adds a way to validate parsed configs in `SparkConfParser`. It is intended for the following use:

```
public int parquetBatchSize() {
  return confParser
      .intConf()
      ...
      .check(value -> value > 0, "Parquet batch size must be > 0")
      .parse();
}
```

### CODE REVIEW

File: spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/SparkConfParser.java [hunk 1]
```diff
@@ -227,7 +225,26 @@ public ThisT tableProperty(String name) {
...
     }
 
-    protected T parse(Function<String, T> conversion, T defaultValue) {
+    public ThisT check(Predicate<T> newCheck, String newErrMsg) {
+      this.check = newCheck;
+      this.errMsg = newErrMsg;
+      return self();
+    }
+
+    protected T parseAndCheck(Function<String, T> conversion, T defaultValue) {
```
Comment (manuzhang) [inline spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/SparkConfParser.java:234]: Is it possible for `ProcedureInput` to utilize this method as well?

Comment (aokolnychyi) [reply spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/SparkConfParser.java:234]: This one depends on some state, I'll have to take a look `ProcedureInput`.

File: spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/SparkConfParser.java [hunk 2]
```diff
@@ -264,7 +261,26 @@ public ThisT tableProperty(String name) {
...
+      T value = parse(conversion, defaultValue);
+      Preconditions.checkArgument(value != null, "Value cannot be null");
+      check(value);
+      return value;
+    }
+
+    protected T parseOptionalAndCheck(Function<String, T> conversion, T defaultValue) {
+      T value = parse(conversion, defaultValue);
```
Comment (szehon-ho) [inline spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/SparkConfParser.java:278]: Why do we need a new method? In the existing parse() methods, if check != null can be the case to trigger the check?

Comment (aokolnychyi) [reply spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/SparkConfParser.java:278]: You mean what's the benefit of having `parseOptionalAndCheck` if we can simply call `check` in children?

Comment (szehon-ho) [reply spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/SparkConfParser.java:278]: Oh I mean like adding in existing base ConfParser.parse(Function<String, T> conversion, T defaultValue) method

```
if (check != null) {
   check(value);
}
```

File: spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/SparkConfParser.java [hunk 3]
```diff
@@ -233,19 +229,20 @@ public EnumConfParser<T> defaultValue(String value) {
...
-      return parse(toEnum, defaultValue);
+      return parseOptionalAndCheck(toEnum, defaultValue);
     }
   }
 
   abstract class ConfParser<ThisT, T> {
     private final List<String> optionNames = Lists.newArrayList();
     private String sessionConfName;
     private String tablePropertyName;
+    private Predicate<T> check;
```
Comment (jerqi) [inline spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/SparkConfParser.java:244]: Maybe validator is a better name.

File: spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/SparkConfParser.java [hunk 4]
```diff
@@ -264,7 +261,26 @@ public ThisT tableProperty(String name) {
...
+    protected T parseAndCheck(Function<String, T> conversion, T defaultValue) {
+      T value = parse(conversion, defaultValue);
+      Preconditions.checkArgument(value != null, "Value cannot be null");
+      check(value);
+      return value;
+    }
+
+    protected T parseOptionalAndCheck(Function<String, T> conversion, T defaultValue) {
+      T value = parse(conversion, defaultValue);
+      check(value);
```
Comment (jerqi) [inline spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/SparkConfParser.java:279]: Should we skip the check if this value is null for optional config?

**Pull request closed without merging**